sql: handle UPDATEs for partial indexes

mgartner opened the pull request:
#### sql: handle UPDATEs for partial indexes

This commit enhances the behavoir of `UPDATE` so that the state of
partial indexes is correctly maintained. Two boolean columns are
synthesized for each partial index defined on the table. These two
columns indicate whether the old rows and new rows satisfy the partial
index predicate, respectively. During execution, these columns are
accessed when deciding whether to delete or put entries in the partial
index.

Note that `UPDATE`s to partial inverted indexes are not yet supported or
tested.

Fixes #50221

Release note: None

#### sql: shortcut logic in mutations for tables without partial indexes

This commit avoids a loop through a table's indexes for every row
processed in a mutation when the table contains no partial indexes. By
caching the set of ordinals of the partial indexes defined on the table,
we can avoid partial-index related logic in the case that the set is
empty. Also, this set allows iterating through just the partial indexes,
not all the indexes.

Release note: None

mgartner committed:
sql: handle UPDATEs for partial indexes

This commit enhances the behavoir of `UPDATE` so that the state of
partial indexes is correctly maintained. Two boolean columns are
synthesized for each partial index defined on the table. These two
columns indicate whether the old rows and new rows satisfy the partial
index predicate, respectively. During execution, these columns are
accessed when deciding whether to delete or put entries in the partial
index.

Note that `UPDATE`s to partial inverted indexes are not yet supported or
tested.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50950)
<!-- Reviewable:end -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50950)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/row/updater.go, line 286 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50950#-MBH2fiuEo835LNztY7s:-MBH2fiuEo835LNztY7t:b-tq3map) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb8623835dc5becea6e26feac7acb30047c932b3/pkg/sql/row/updater.go#L286)):*
> ```
> 		// TODO(mgartner): Add partial index IDs to ignoreIndexes that we should
> 		// not delete entries from.
> ```

I'm not entirely sure whether or not we should be ignoring any partial indexes here or not. It sounds like its already making extraneous deletes. 

If we should be ignoring deletes for rows that aren't already in the partial index, how would I go about testing this case?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
Sorry for the delay!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50950#-:-MBjuV3pBXvuLVWwZ2DW:b2c5wy4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[pkg/sql/update.go, line 313 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50950#-MBjpppG7S2XzqDY6Lck:-MBjpppG7S2XzqDY6Lcl:bk2k4on) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb8623835dc5becea6e26feac7acb30047c932b3/pkg/sql/update.go#L313)):*
> ```Go
> 	partialIndexValOffset := len(u.run.tu.ru.FetchCols) + len(u.run.tu.ru.UpdateCols) + u.run.numPassthrough
> 	partialIndexVals := sourceVals[partialIndexValOffset:]
> 	colIdx := 0
> ```

Is there an easy (fast) way to check if there are any partial indexes? It would be good if we could shortcut the loops in the common case of no partial indexes. Maybe we can cache this information in `ImmutableTableDescriptor`?
___
*[pkg/sql/update.go, line 325 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50950#-MBjpYX9BaqW-OFikhgR:-MBjpYX9BaqW-OFikhgS:b-gd5bkt) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb8623835dc5becea6e26feac7acb30047c932b3/pkg/sql/update.go#L325)):*
> ```Go
> 				// If the value of the column for the index predicate expression
> 				// is false, the row should not be added to the partial index.
> 				if i < len(indexes) {
> ```

How can i be >= len(indexes)?
___
*[pkg/sql/update.go, line 336 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50950#-MBjpgcP52J4pFCFcXpJ:-MBjpgcP52J4pFCFcXpK:b-g97kil) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb8623835dc5becea6e26feac7acb30047c932b3/pkg/sql/update.go#L336)):*
> ```Go
> 	for i := range indexes {
> 		index := &indexes[i]
> 		if index.IsPartial() {
> ```

[nit] it would be more succinct to do both in the same loop
___
*[pkg/sql/opt/optbuilder/testdata/update, line 1706 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50950#-MBjrrS9A3iw_0-PjNeQ:-MBjrrS9A3iw_0-PjNeR:br2fm21) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb8623835dc5becea6e26feac7acb30047c932b3/pkg/sql/opt/optbuilder/testdata/update#L1706)):*
> ```
>       │         └── 1 [as=a_new:9]
>       └── projections
>            └── (a_new:9 > b:5) AND (c:6 = 'bar') [as=partial_index_put2:10]
> ```

We should add some testcases with write-only and delete-only partial indexes (I guess this is true for inserts/deletes as well)
___
*[pkg/sql/row/updater.go, line 286 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50950#-MBH2fiuEo835LNztY7s:-MBjrOcu5z4UHYY1bL5_:b-s962ow) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb8623835dc5becea6e26feac7acb30047c932b3/pkg/sql/row/updater.go#L286)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

> ```
> 		// TODO(mgartner): Add partial index IDs to ignoreIndexes that we should
> 		// not delete entries from.
> ```

I'm not entirely sure whether or not we should be ignoring any partial indexes here or not. It sounds like its already making extraneous deletes. 

If we should be ignoring deletes for rows that aren't already in the partial index, how would I go about testing this case?
</blockquote></details>

This seems fine for now.. In principle we would want to check (in the loop at the end of the function) whether the row was supposed to be in that index or not. I don't think this case is common in practice though (I am told it only happens when you add an index and then update the table in the same txn).

BTW it seems that this code could be cleaned up now - we could just put the deleted indexes in `ignoreIndexesForPut` and shouldn't need all this extra logic. Maybe file an issue for that (unfortunately this code has no clear owner).


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50950)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[pkg/sql/update.go, line 325 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50950#-MBjpYX9BaqW-OFikhgR:-MBpBLzs5dJpcRi4Qv9v:be2dg35) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb8623835dc5becea6e26feac7acb30047c932b3/pkg/sql/update.go#L325)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

How can i be >= len(indexes)?
</blockquote></details>

Oops. This was leftover from my first attempt to combine the loops into one. Fixed.
___
*[pkg/sql/update.go, line 336 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50950#-MBjpgcP52J4pFCFcXpJ:-MBpBTqF5zqeWpM0vXc_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb8623835dc5becea6e26feac7acb30047c932b3/pkg/sql/update.go#L336)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] it would be more succinct to do both in the same loop
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mgartner commented:
___
*[pkg/sql/row/updater.go, line 286 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50950#-MBH2fiuEo835LNztY7s:-MBuKzXpE7Qb4FssfTKy:bz2b4i4) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb8623835dc5becea6e26feac7acb30047c932b3/pkg/sql/row/updater.go#L286)):*
I'm all down for the cleanup, but would that be possible. I don't fully understand this comment, but it seems to suggest that we'd have to read all values of the updated rows in order to simplify this.

> // We choose to make this compromise in order to avoid
> // having to read all values of the row that is being updated.


<!-- Sent from Reviewable.io -->

mgartner committed:
sql: shortcut logic in mutations for tables without partial indexes

This commit avoids a loop through a table's indexes for every row
processed in a mutation when the table contains no partial indexes. By
caching the set of ordinals of the partial indexes defined on the table,
we can avoid partial-index related logic in the case that the set is
empty. Also, this set allows iterating through just the partial indexes,
not all the indexes.

Release note: None

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50950)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[pkg/sql/update.go, line 313 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50950#-MBjpppG7S2XzqDY6Lck:-MBv7LtV9LXzvAAfwIH0:bfonp5l) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb8623835dc5becea6e26feac7acb30047c932b3/pkg/sql/update.go#L313)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Is there an easy (fast) way to check if there are any partial indexes? It would be good if we could shortcut the loops in the common case of no partial indexes. Maybe we can cache this information in `ImmutableTableDescriptor`?
</blockquote></details>

I've added a set that includes the partial index ordinals to `ImmutableTableDescriptor` and used it for inserts, delete and updates. The advantage of the ordinals is that it can iterate through just the partial indexes. If you think a boolean is sufficient, I can change it.
___
*[pkg/sql/opt/optbuilder/testdata/update, line 1706 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50950#-MBjrrS9A3iw_0-PjNeQ:-MBuKYjU4DaEvRj9nrxz:bga7dig) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb8623835dc5becea6e26feac7acb30047c932b3/pkg/sql/opt/optbuilder/testdata/update#L1706)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We should add some testcases with write-only and delete-only partial indexes (I guess this is true for inserts/deletes as well)
</blockquote></details>

Done. Looks like that was a bug—there were no synthesized columns for write-only and delete-only indexes.


<!-- Sent from Reviewable.io -->

mgartner commented on pkg/sql/update.go:
```diff
@@ -305,8 +306,49 @@ func (u *updateNode) processSourceRow(params runParams, sourceVals tree.Datums)
+	var ignoreIndexesForDel util.FastIntSet
+	partialIndexOrds := u.run.tu.tableDesc().PartialIndexOrds()
+	if !partialIndexOrds.Empty() {
+		partialIndexValOffset := len(u.run.tu.ru.FetchCols) + len(u.run.tu.ru.UpdateCols) + u.run.numPassthrough
```
This looks like a bug. I need to add `u.run.checkOrds.Len()`.

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50950#-:-MCEKPHqByPSqtZW1IGQ:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)

___
*[pkg/sql/update.go, line 315 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50950#-MBH2WGSdtGVV7blniQ--r3-315:-MCEKhavCDwQKBivoIch:bc5fqo3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6de87a313733df1eafd83c831c7aaf4b727b0178/pkg/sql/update.go#L315)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

This looks like a bug. I need to add `u.run.checkOrds.Len()`.
</blockquote></details>

Hm, maybe add a check constraint in the table in the partial_index test


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50950)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/update.go, line 315 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50950#-MBH2WGSdtGVV7blniQ--r3-315:-MCEYGOXECIR2PpNqUW1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6de87a313733df1eafd83c831c7aaf4b727b0178/pkg/sql/update.go#L315)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Hm, maybe add a check constraint in the table in the partial_index test
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mgartner commented:
bors r=RaduBerinde
